server/mmoauth/mmoauth.go

```diff
@@ -0,0 +1,248 @@
+		return
+	}
+
+	redirectCookie, err := r.Cookie("oauthredirect")
+	if err != nil {
+		a.handleError("oauthredirect cookie not found", err, w, r)
+		return
+	}
+
+	// Delete cookies
+	deleteCookie("oauthredirect", w)
+	deleteCookie("oauthstate", w)
+
+	if r.FormValue("state") != stateCookie.Value {
+		log.Println("invalid oauth state")
+		http.Redirect(w, r, "/", http.StatusTemporaryRedirect)
+		return
+	}
+
+	code := r.FormValue("code")
+	token, err := a.config.Exchange(context.Background(), code)
+	if err != nil {
+		a.handleError("code exchange error", err, w, r)
+		return
+	}
+
+	// Get user info from Mattermost
+	client := mattermost.NewAPIv4Client(a.params.MattermostURL)
+	client.SetOAuthToken(token.AccessToken)
+
+	etag := ""
+	user, resp := client.GetMe(etag)
+	if resp.Error != nil {
+		a.handleError("failed getting Mattermost user", resp.Error, w, r)
+		return
+	}
+
+	log.Printf("UserInfo: %v\n", user)
+
+	teams, resp := client.GetAllTeams(etag, 0, 10)
+	if resp.Error != nil {
+		a.handleError("failed getting Mattermost teams", err, w, r)
+		return
+	}
+
+	if len(teams) < 1 {
+		a.handleError("User has no teams", err, w, r)
+		return
+	}
+
+	log.Printf("%d team(s).", len(teams))
+	teamId := teams[0].Id
+
+	channels, resp := client.GetChannelsForTeamForUser(teamId, user.Id, false, etag)
+	if resp.Error != nil {
+		a.handleError("failed getting Mattermost channels for user", err, w, r)
+		return
+	}
+
+	if len(channels) < 1 {
+		a.handleError("User has no channels", err, w, r)
+		return
+	}
+
+	log.Printf("%d channel(s).", len(channels))
+	channelId := channels[0].Id
+
+	log.Printf("userId: %s, teamId: %s, channelId: %s", user.Id, teamId, channelId)
+
+	// Create native user
+	nativeUser, _ := a.store.GetUserById(user.Id)
+	if nativeUser == nil {
+		err = a.store.CreateUser(&model.User{
+			ID:          user.Id,
+			Username:    user.Username,
+			Email:       user.Email,
+			Password:    "",
+			MfaSecret:   "",
+			AuthService: "mattermost",
+			AuthData:    "",
+			Props:       map[string]interface{}{},
+		})
+		if err != nil {
+			a.handleError("Unable to create the new user", err, w, r)
+			return
+		}
+	}
+
+	// Create session
+	// TODO: Remove any existing sessions with the same token? (Happens on re-login)
+	session := model.Session{
+		ID:     uuid.New().String(),
+		Token:  token.AccessToken,
+		UserID: user.Id,
+		Props: map[string]interface{}{
+			"authService":  "mattermost",
+			"expiry":       token.Expiry.Unix(),
+			"refreshToken": token.RefreshToken,
+		},
+	}
+	err = a.store.CreateSession(&session)
+	if err != nil {
+		a.handleError("unable to create session", err, w, r)
+		return
+	}
+
+	// Pass access token to client via cookie
+	// TODO: Review security of this approach
+	var expiration = time.Now().Add(365 * 24 * time.Hour)
+	tokenCookie := http.Cookie{Name: "oauthtoken", Path: "/", Value: token.AccessToken, Expires: expiration, HttpOnly: false}
+	http.SetCookie(w, &tokenCookie)
```

Comment:
I don't think this part needs to be specific to an oauth token. I would unify this with how non-oauth login works (which looks like via localstorage). Having two different ways of storing the authentication token on the webapp/client increase the attack surface and chance for bad security bugs IMO. And the webapp/client doesn't really need to care about whether it's oauth or not once it has it's session.

Reply:
So, the `oauthtoken` cookie is just used to pass it to the client JS code, which (in app.tsx) copies it into localStorage and deletes it. This is because OAuth does a redirect to the callback, instead of going through a fetch like the login API. Do you, or does anyone else know of a better way to pass the token back to the client code?

Reply:
Got it. Missed that. One way to do it is to have the redirect go to a webapp page that makes the API request to the oauth callback endpoint. That way you could still return a JSON as a response for the webapp to handle the token from like you do for regular login. Other option would be to have regular login use a cookie too. WDYT?

Reply:
Thanks for the idea! I realized the callback page can just render JS to set the token and redirect, so reimplemented it that way and removed the unnecessary 'oauthtoken' cookie.